### ISSUE TITLE
Multiple windows aren't properly resized on Wayland

### ISSUE THREAD

**Issue by kchibisov** [alacritty/alacritty#6069]
In some cases the old buffer could be submitted. We do resize everything, issue the call to EGL, however for whatever reason the dmabuf is still having old sizes until some draw command in the future(not the next one, since the next one after resize failed).

I've used `wlhax` to debug it and there's a clear notion of old buffer and messed up state due to mesa, since `glViewport` is also wrong and relative to the old window, and not the current one. Which could only mean that for whatever reason resize failed.

**Comment by mahkoh** [comment]
I just encountered a similar bug though I don't know if it is exactly what you are seeing. If the alacritty window is not focused, then, after resizing, the new area is not filled in by alacritty. This only happens when the window is resized quickly.

I have traced this bug to the following lines: https://github.com/alacritty/alacritty/blob/56a69c0bfe5742aeda1c4942b748f5a842ce6773/alacritty/src/window_context.rs#L321-L325

This flag is set in https://github.com/alacritty/alacritty/blob/56a69c0bfe5742aeda1c4942b748f5a842ce6773/alacritty/src/display/mod.rs#L1109-L1111

However, there is nothing in that callback that would trigger alacritty to actually do any drawing. I believe that, in this callback, there should be something that triggers glutin to re-emit the `RedrawEventsCleared`, e.g. a call to `request_redraw`. But I'm not sure how to best implement this.

---

Now, why this only happens in multi-window mode I don't know.

**Comment by mahkoh** [comment]
This might actually be a bug in mesa.

1. make_current causes mesa to allocate buffers for the current window size
2. If buffers are allocated, wl_egl_window_resize has effect until the next swap call.
3. alacritty calls make_current and wl_egl_window_resize multiple times without intermediate swap calls.

This would explain why, on the first swap call afterwards, you're seeing buffers allocated with the size as it was at the time of the first  wl_egl_window_resize call. Only after this swap call does the last wl_egl_window_resize call take effect.

**Comment by kchibisov** [comment]
This what I think is happening as well and thats what mesa dubugging showed me. We can resize before making current though as a work around. Since resize doesn't rrquire it.

**Comment by mahkoh** [comment]
I tried this but unfortunately the problem was still present.

**Comment by mahkoh** [comment]
As a workaround I also tried simply filling the current buffer with the background color and then swapping immediately before the resize. This works consistently but unfortunately you will get a blank window while resizing.

**Comment by kchibisov** [comment]
@mahkoh I was trying to start https://gitlab.freedesktop.org/mesa/mesa/-/issues/6547 to figure it out.

Though, it was marked as works as expected.

**Comment by kchibisov** [comment]
@mahkoh 

Have you tried
```diff
diff --git a/alacritty/src/display/mod.rs b/alacritty/src/display/mod.rs
index 91f575e4..4e9a8bf6 100644
--- a/alacritty/src/display/mod.rs
+++ b/alacritty/src/display/mod.rs
@@ -568,9 +568,6 @@ impl Display {
         let (mut cell_width, mut cell_height) =
             (self.size_info.cell_width(), self.size_info.cell_height());
 
-        // Ensure we're modifying the correct OpenGL context.
-        self.window.make_current();
-
         // Update font size and cell dimensions.
         if let Some(font) = pending_update.font() {
             let scale_factor = self.window.scale_factor;
@@ -622,6 +619,10 @@ impl Display {
         // Resize renderer.
         let physical = PhysicalSize::new(self.size_info.width() as _, self.size_info.height() as _);
         self.window.resize(physical);
+
+        // Ensure we're modifying the correct OpenGL context.
+        self.window.make_current();
+
         self.renderer.resize(&self.size_info);
 
         if self.collect_damage() {
```

**Comment by mahkoh** [comment]
Yes.


https://user-images.githubusercontent.com/1882250/171935917-28aafe07-04be-454e-82c9-e162aa0f3626.mp4

**Comment by kchibisov** [comment]
@mahkoh if you combine it with https://github.com/alacritty/alacritty/pull/6079 does it fix the issue?

**Comment by kchibisov** [comment]
Don't get me wrong, I do think that it's something up with mesa, but at least getting it worked around isn't that bad, since it's a video driver...

**Comment by mahkoh** [comment]
Same result.

**Comment by mahkoh** [comment]
The one thing that I find resolves the problem is to remove https://github.com/alacritty/alacritty/blob/56a69c0bfe5742aeda1c4942b748f5a842ce6773/alacritty/src/window_context.rs#L321-L325

**Comment by kchibisov** [comment]
Hm, if that does resolve then I think wakeing up winit could work with the combination of moving `make_current`.

I'll try to write something. thx.

**Comment by mahkoh** [comment]
```diff
diff --git a/alacritty/src/display/mod.rs b/alacritty/src/display/mod.rs
index 91f575e4..695d3631 100644
--- a/alacritty/src/display/mod.rs
+++ b/alacritty/src/display/mod.rs
@@ -8,7 +8,7 @@ use std::{cmp, mem};
 
 use glutin::dpi::PhysicalSize;
 use glutin::event::ModifiersState;
-use glutin::event_loop::EventLoopWindowTarget;
+use glutin::event_loop::{EventLoopProxy, EventLoopWindowTarget};
 #[cfg(all(feature = "x11", not(any(target_os = "macos", windows))))]
 use glutin::platform::unix::EventLoopWindowTargetExtUnix;
 use glutin::window::CursorIcon;
@@ -45,10 +45,11 @@ use crate::display::damage::RenderDamageIterator;
 use crate::display::hint::{HintMatch, HintState};
 use crate::display::meter::Meter;
 use crate::display::window::Window;
-use crate::event::{Mouse, SearchState};
+use crate::event::{EventType, Mouse, SearchState};
 use crate::message_bar::{MessageBuffer, MessageType};
 use crate::renderer::rects::{RenderLines, RenderRect};
 use crate::renderer::{self, GlyphCache, Renderer};
+use crate::Event;
 
 pub mod content;
 pub mod cursor;
@@ -694,6 +695,7 @@ impl Display {
     pub fn draw<T: EventListener>(
         &mut self,
         mut terminal: MutexGuard<'_, Term<T>>,
+        event_proxy: &EventLoopProxy<Event>,
         message_buffer: &MessageBuffer,
         config: &UiConfig,
         search_state: &SearchState,
@@ -861,7 +863,7 @@ impl Display {
         // Frame event should be requested before swaping buffers, since it requires surface
         // `commit`, which is done by swap buffers under the hood.
         #[cfg(all(feature = "wayland", not(any(target_os = "macos", windows))))]
-        self.request_frame(&self.window);
+        self.request_frame(&self.window, event_proxy);
 
         // Clearing debug highlights from the previous frame requires full redraw.
         if self.is_damage_supported && !self.debug_damage {
@@ -1094,7 +1096,7 @@ impl Display {
     /// Requst a new frame for a window on Wayland.
     #[inline]
     #[cfg(all(feature = "wayland", not(any(target_os = "macos", windows))))]
-    fn request_frame(&self, window: &Window) {
+    fn request_frame(&self, window: &Window, event_proxy: &EventLoopProxy<Event>) {
         let surface = match window.wayland_surface() {
             Some(surface) => surface,
             None => return,
@@ -1106,8 +1108,11 @@ impl Display {
         should_draw.store(false, Ordering::Relaxed);
 
         // Request a new frame.
+        let proxy = event_proxy.clone();
+        let id = window.id();
         surface.frame().quick_assign(move |_, _, _| {
             should_draw.store(true, Ordering::Relaxed);
+            let _ = proxy.send_event(Event::new(EventType::Redraw, id));
         });
     }
 }
diff --git a/alacritty/src/event.rs b/alacritty/src/event.rs
index d16114af..c9427a50 100644
--- a/alacritty/src/event.rs
+++ b/alacritty/src/event.rs
@@ -91,6 +91,7 @@ pub enum EventType {
     CreateWindow(WindowOptions),
     BlinkCursor,
     SearchNext,
+    Redraw,
 }
 
 impl From<TerminalEvent> for EventType {
@@ -1107,6 +1108,9 @@ impl input::Processor<EventProxy, ActionContext<'_, Notifier, EventProxy>> {
                     TerminalEvent::CursorBlinkingChange => self.ctx.update_cursor_blinking(),
                 },
                 EventType::ConfigReload(_) | EventType::CreateWindow(_) => (),
+                EventType::Redraw => {
+                    *self.ctx.dirty = true
+                },
             },
             GlutinEvent::RedrawRequested(_) => *self.ctx.dirty = true,
             GlutinEvent::WindowEvent { event, .. } => {
diff --git a/alacritty/src/window_context.rs b/alacritty/src/window_context.rs
index 6ead6e4e..15dbd261 100644
--- a/alacritty/src/window_context.rs
+++ b/alacritty/src/window_context.rs
@@ -333,7 +333,13 @@ impl WindowContext {
             }
 
             // Redraw screen.
-            self.display.draw(terminal, &self.message_buffer, config, &self.search_state);
+            self.display.draw(
+                terminal,
+                event_proxy,
+                &self.message_buffer,
+                config,
+                &self.search_state,
+            );
         }
     }
 
```
works for me

**Comment by kchibisov** [comment]
I wonder if we can send something that does just `RedrawEventsCleared` + Move the `current_context` logic after window resize but before renderer resize?

So we won't have that dirty, but will have an event to wakeup the event loop in advance. Though, I'd have to test your logic a bit to understand whether it fixes or not. It's just in your particular issue the problem was that I think we've already done processing events for the rest of the windows for which the frame callback got triggered, so it doesn't woke up properly in your particular case, just forwarding event should do that + we should ensure that the buffer is new, which we can do by moving context related stuff in `handle_update`. I'll get to it tomorrow though, but thanks for your input, it does help a lot.

**Comment by kchibisov** [comment]
@mahkoh does this works #6109?

**Comment by mahkoh** [comment]
Same result.

**Comment by kchibisov** [comment]
@mahkoh the issue is that your patch doesn't fix my issue, but mine doesn't fix yourth.

To fix my issue I have to move `make_current` stuff after `window.resize`, once I do this it vanishes. So Let me try to force dirty bit...

**Comment by mahkoh** [comment]
Interesting. Could you post a video of what you're seeing?

For reference: I'm using an AMD graphics card. How the bug manifests might also depend on the compositor you're using and the order/frequency in which it emits events. I'm using my own compositor which it significantly more aggressive than other wayland compositors when it comes to the frequency in which it emits resize events. This has also uncovered bugs in other programs in the past that were not reproducible on other compositors: https://github.com/mpv-player/mpv/pull/9813

**Comment by kchibisov** [comment]
@mahkoh I've pushed one more commit, could you test it?

**Comment by kchibisov** [comment]
@mahkoh actually, do you see an infinite wakeup loop with your patch? Since that's what I'm observing and that's something I'd expect, given that sending event via proxy will enqueue for the next frame, however frame callback will be handled from the redraw events cleared resulting in processing `dirty` before.

So I'm not entirely sure how that should be handled, well we can remove dirty bit when sending event on the terminal and wait for it.

**Comment by mahkoh** [comment]
As you've also discovered, my initial patch only hides the issue because it causes alacritty to redraw at the display refresh rate.

Unfortunately, your latest patch, while it prevents these redraws, does not fix the bug. I've added some printfs to further debug the issue combined with WAYLAND_DEBUG=1. The following happens when I rapidly resize two windows. `RESIZE` is printed immediately before window.resize, `DRAW` is printed immediately before a draw. 15 refers to the window that fails to resize. 36 refers to the window that resizes correctly. 

```
RESIZE 15 PhysicalSize { width: 459, height: 521 }
RESIZE 36 PhysicalSize { width: 497, height: 521 }
RESIZE 15 PhysicalSize { width: 478, height: 521 }
RESIZE 36 PhysicalSize { width: 478, height: 521 }
RESIZE 15 PhysicalSize { width: 495, height: 521 }
RESIZE 36 PhysicalSize { width: 461, height: 521 }
RESIZE 15 PhysicalSize { width: 515, height: 521 }
RESIZE 36 PhysicalSize { width: 441, height: 521 }
RESIZE 15 PhysicalSize { width: 534, height: 521 }
RESIZE 36 PhysicalSize { width: 422, height: 521 }
DRAW 15
[2001162.448]  -> zwp_linux_buffer_params_v1@47.create_immed(new id wl_buffer@48, 459, 521, 875713089, 0)
[2001162.471]  -> wl_surface@15.attach(wl_buffer@48, 0, 0)
DRAW 36
[2001164.085]  -> zwp_linux_buffer_params_v1@46.create_immed(new id wl_buffer@50, 497, 521, 875713089, 0)
[2001164.108]  -> wl_surface@36.attach(wl_buffer@50, 0, 0)
DRAW 36
[2001213.269]  -> zwp_linux_buffer_params_v1@52.create_immed(new id wl_buffer@54, 422, 521, 875713089, 0)
[2001213.289]  -> wl_surface@36.attach(wl_buffer@54, 0, 0)
```

I believe 36 redrawing twice at the end is just a fluke due to it having the keyboard focus.

At this point I see no way around doing two draws and two swaps after each resize. I.e., if we have not drawn since the last resize, call `draw` twice to work around this issue.

**Comment by kchibisov** [comment]
@mahkoh are you testing on sway or your compositor? Since I can't repro your particular issue under sway.

**Comment by mahkoh** [comment]
I'm testing under my compositor. Btw I've figured out why moving make_current did not work: Even if we only call make_current after the resize, subsequent resizes will still not take effect since we already called make_current once. I suppose as a workaround all of this work requiring make_current could be done lazily in `draw` itself.

**Comment by kchibisov** [comment]
Hm, I think I start understanding what is happening.

Winit can send only one resize per event loop tick, however we're not getting frame callback at this point, so we do `make_current`, but we don't draw, hence no frame. So it works for me since either my compositor is fast enough or I do single resize.

Let me try one more thing.
